File: 2024/pkg/day-one/day-one.go
package dayone

import (
	"fmt"
	"log/slog"
	"slices"
	"strconv"
	"strings"

	"github.com/skykosiner/AOC/pkg/utils"
)

type Item struct {
	value int
	seen  int
}

func (i Item) String() string {
	return fmt.Sprintf("%d %d\n", i.value, i.seen)
}

func leftRightsides(lines []string) ([]int, []int) {
	var left []int
	var right []int

	for _, line := range lines {
		sides := strings.Split(line, "   ")
		if len(sides) == 1 {
			break
		}

		l, _ := strconv.Atoi(sides[0])
		r, _ := strconv.Atoi(sides[1])
		left = append(left, l)
		right = append(right, r)
	}

	return left, right
}

func PartOne() {
	var distances []int

	lines, err := utils.ReadFile("./pkg/day-one/input")
	if err != nil {
		slog.Error("Error reading input file.", "error", err)
		return
	}

	left, right := leftRightsides(lines)

	slices.Sort(left)
	slices.Sort(right)

	for idx := range left {
		l := left[idx]
		r := right[idx]

		if l > r {
			distances = append(distances, int(left[idx]-right[idx]))
		} else {
			distances = append(distances, int(right[idx]-left[idx]))
		}
	}

	fmt.Println(utils.Sum(distances))
}

func PartTwo() {
	var items []Item
	var similarities []int

	lines, err := utils.ReadFile("./pkg/day-one/input")
	if err != nil {
		slog.Error("Error reading input file.", "error", err)
		return
	}

	left, right := leftRightsides(lines)

	for _, l := range left {
		seen := 0
		for _, r := range right {
			if l == r {
				seen++
			}
		}

		items = append(items, Item{
			value: l,
			seen:  seen,
		})
	}

	for _, item := range items {
		similarities = append(similarities, item.seen*item.value)
	}

	fmt.Println(utils.Sum(similarities))
}

File: 2024/src/day-one.ts
import fs from "fs";

function splitLeftAndRight(lines: string[]): [number[], number[]] {
    const left: number[] = [];
    const right: number[] = [];

    for (let i = 0; i < lines.length; i++) {
        const nums = lines[i].split("   ");
        if (nums.length === 1) {
            break;
        }

        left.push(+nums[0]);
        right.push(+nums[1]);
    };

    return [left, right]
}

const lines = fs.readFileSync("../day1/input").toString().split("\n");
const [left, right] = splitLeftAndRight(lines);
left.sort()
right.sort()

let distance = 0;
for (let i = 0; i < left.length; i++) {
    distance += Math.abs(left[i] - right[i])
}

console.log(distance);
